redis/src/main/java/org/hibernate/ogm/datastore/redis/RedisHashDialect.java

```diff
@@ -0,0 +1,282 @@
+
+/**
+* Stores tuples and associations inside Redis using hash data structures.
+ * <p>
+ * Tuples are stored in Redis within hashes. Associations are stored in Redis obtained as either single values or a
+ * JSON serialization of a {@link org.hibernate.ogm.datastore.redis.dialect.value.Association} object in list data structures.
+ *
+ * @author Mark Paluch
+ */
+public class RedisHashDialect extends AbstractRedisDialect {
+
+	public RedisHashDialect(RedisDatastoreProvider provider) {
+		super( provider.getConnection() );
+	}
+
+	@Override
+	public GridType overrideType(Type type) {
+		return RedisHashTypeConverter.INSTANCE.convert( type );
+	}
+
+	@Override
+	@SuppressWarnings({ "unchecked", "rawtypes" })
+	public Tuple getTuple(
+			EntityKey key, TupleContext tupleContext) {
+		String entityIdString = entityId( key );
+		if ( !connection.exists( entityIdString ) ) {
+			return null;
+		}
+
+		Map<String, Object> objects;
+		if ( tupleContext.getSelectableColumns().isEmpty() ) {
+			objects = (Map) connection.hgetall( entityIdString );
+		}
+		else {
+			List<String> hmget = connection.hmget( entityIdString, getFields( tupleContext ) );
+			objects = toEntity( tupleContext, hmget );
+		}
+
+		return new Tuple( new RedisTupleSnapshot( objects ) );
+	}
+
+	private Map<String, Object> toEntity(TupleContext tupleContext, List<String> hmget) {
+		Map<String, Object> objects = new HashMap<>();
+		for ( int i = 0; i < tupleContext.getSelectableColumns().size(); i++ ) {
+			String columnName = tupleContext.getSelectableColumns().get( i );
+			String value = hmget.get( i );
+			if ( value == null ) {
+				continue;
+			}
+			objects.put( columnName, value );
+		}
+		return objects;
+	}
+
+	private String[] getFields(TupleContext tupleContext) {
+		return tupleContext.getSelectableColumns().toArray( new String[tupleContext.getSelectableColumns().size()] );
+	}
+
+	@Override
+	public void insertOrUpdateTuple(
+			EntityKey key, Tuple tuple, TupleContext tupleContext) throws TupleAlreadyExistsException {
+		Map<String, Object> map = ( (RedisTupleSnapshot) tuple.getSnapshot() ).getMap();
+		Map<String, String> entity = new HashMap<>();
+		List<String> toDelete1 = new ArrayList<>();
+
+		for ( TupleOperation action : tuple.getOperations() ) {
+			switch ( action.getType() ) {
+				case PUT:
+					map.put( action.getColumn(), action.getValue() );
+
+					// TODO: IllegalStateException when value is not String/Character?
+					if ( action.getValue() instanceof Character ) {
+						entity.put( action.getColumn(), action.getValue().toString() );
+					}
+					else {
+						entity.put( action.getColumn(), (String) action.getValue() );
+					}
+					break;
+				case REMOVE:
+				case PUT_NULL:
+					map.remove( action.getColumn() );
+					toDelete1.add( action.getColumn() );
+					break;
+			}
+		}
+
+		String entityId = entityId( key );
+		Long currentTtl = connection.pttl( entityId( key ) );
+
+		if ( !toDelete1.isEmpty() ) {
```

Comment:
The numbered variable names are a bit tough to grok. Maybe you can find something more descritive? Or just `toDelete` and inline the second one.